### ISSUE TITLE
CloudWatch: grafana 8.4.0 - Request cancelled: context deadline exceeded

### ISSUE THREAD

**Issue by salavessa** [grafana/grafana#45563]
**What happened**:
After updating from `8.3.6` to `8.4.0` servers started to trigger alerts with error `request handler response error {metric request error: "RequestCanceled: request context canceled\ncaused by: context deadline exceeded" <nil> [] [] 0xc001ca3140}`.
It was also observed timeouts or general performance degradation while loading _Dashboard Variables_ initialized via _CloudWatch_ queries.

**What you expected to happen**:
No errors.

**How to reproduce it (as minimally and precisely as possible)**:
Configuring a dashboard with multiple _CloudWatch_ queries/datasources and eventually some Dashboar Variables from _CloudWatch_ datasources,

**Anything else we need to know?**:
The exact same issue happened in three different servers, fetching from different AWS accounts/regions and asset types.
Reverted to `8.3.6` on all servers and everything started to work fine again.

**Environment**:
- Grafana version: 8.4.0
- Data source type & version: CloudWatch
- OS Grafana is installed on: linux (docker)
- User OS & Browser: win10 & chrome 98
- Grafana plugins: n/a
- Others: official grafana docker image

**Comment by zuchka** [comment]
Thanks for creating this issue, @salavessa 

Can you provide any more details about this issue? I just viewed some of our internals dashboards that use Cloudwatch queries with template variables and they seem to work. That instance is running `8.5.0-pre`.

If you turn your logging level up to `debug`, does the grafana server logs  offer any more details? 👍

**Comment by salavessa** [comment]
Thanks for coming back to this @zuchka

I couldn't really find anything relevant in `debug` but playing a bit more around it feels that the issue comes from some sort of "concurrency lock" which gets triggered, e.g. when having alerts from `CloudWatch` running in the background while also loading Variables from `CloudWatch` queries via UI.
Maybe also worth mentioning that I'm not using the new grafana's unified alerting mechanism, i.e. `GF_UNIFIED_ALERTING_ENABLED=false`.

I'm able to replicate this behavior quite easily using the steps below (tested on `8.4.1` and `8.5.0-53082pre (0e7b0f16b8)`):
1. Configure a dashboard with multiple Variables whith values populated by `CloudWatch` queries (no need to add any panels) - I'm using 4 variables, getting data from two regions for two different AWS accounts and the queries look like `dimension_values(<REGION>,AWS/ApplicationELB,ConsumedLCUs,LoadBalancer)`
1. Create alerts for `CloudWatch` queries (I have these in a different dashboard) - Evaluate every `1m` For `0m` - I tested with 3 alerts enabled on a `Graph (old)` panel using a query like `SEARCH('{AWS/ApplicationELB,TargetGroup,LoadBalancer} MetricName="UnHealthyHostCount"', 'Maximum', 60)` and having the alert being triggered `WHEN last() query(A, 5m, now) IS ABOVE 0`
1. now we just need to start reloading the dashboard which contains the `CloudWatch Variables` via browser (e.g. `F5` or `Ctrl+R` keys) and, at some point, the page will stay in the `Services spinner` for long time and eventually alerts with errors such as `request handler response error {metric request error: "RequestCanceled: request context canceled\ncaused by: context deadline exceeded" <nil> [] [] 0xc00128a270}` will also start being triggered


**MORE:** Some extra findings which seem to be related with this issue and may make the debugging process a bit simpler:
- Immediately after restarting grafana's container (tested with `8.4.0`, `8.4.1` and `8.5.0-53082pre`), and having no alerts configured (I actually deleted all alerts and dashboards except the one with `CoudWatch` variables only), it seems that the "concurrency lock" also kicks in during startup as it takes around 2 minutes for the `CloudWatch Variables only` dashboard to get over the `Services` spinner/loader
- After seeing the behavior above consistently, I rolled-back grafana container to version `8.3.6` and that same dashboard consistently loads in less than 5 seconds after container restarts


Hope it helps.

Thanks!

**Comment by zuchka** [comment]
awesome research, @salavessa -- TY!

**Comment by yaelleC** [comment]
Thanks for the very detailed bug report @salavessa !

I am trying to replicate your issue, I'll keep you updated as soon as I find anything 🙂

**Comment by marefr** [comment]
Might be related to #44036, but not sure

**Comment by yaelleC** [comment]
Hi @salavessa , sorry for taking so long, we've had issues with panel alerts and permissions and we had to fix a couple of bugs before I could try and replicate.

Unfortunately I was not able to replicate your issue though the dashboard always loads for me... I will try again with different settings, but in the meantime if you wanted to checked whether @marefr 's guess could be the culprit, you can try the images on : https://hub.docker.com/repository/docker/yaelle/grafana

yaelle/grafana:fffd4c10b75bb36c031bc9c6c518463a75cc1963 correspond to the code before #44036 and yaelle/grafana:4e1c7e868eab0fa28903d847ac21628b96297118 the one after. If the first one works and the second does not, we can look into it in more details.

**Comment by salavessa** [comment]
Hi @yaelleC, thanks for this.
Unfortunately I'm running in `linux/amd64` and the images provided seem to be for `linux/arm64/v8`.
Could you please provide images for `linux/amd64` instead?
Thanks!

**Comment by yaelleC** [comment]
Hi @salavessa , you can try again: I made an `after` and `before` tags images that should be for the right platform this time 😉

**Comment by salavessa** [comment]
Hi @yaelleC thanks for those.
I just finished checking this and effectively the issue seems to be related to #44036.
`yaelle/grafana:before` is working fine but with `yaelle/grafana:after` I'm experiencing the issues reported before.
Thanks!

**Comment by yaelleC** [comment]
Thanks very much for confirming that. 
I have added the bug to our backlog, we'll keep you updated 🙂

**Comment by marefr** [comment]
Can you please try configure [timeout](https://grafana.com/docs/grafana/latest/administration/configuration/#timeout) to something more suitable for your queries since I imagine they take longer than 30 seconds?

**Comment by salavessa** [comment]
Hi @marefr.

As mentioned in a [previous comment](https://github.com/grafana/grafana/issues/45563#issuecomment-1049930309), the amount time the "lock" is in place while loading the dashboard after a container restart was pretty much constant (2 minutes).

After testing `8.4.3` with different `GF_DATAPROXY_TIMEOUT` values (`30` - default, `60`, `10`, `5`, `2`) I observed that there seems to be some sort of deadlock around it. I tested refreshing dashboards with different number of variables populated from `CloudWatch` queries and it didn't seem to make any difference the number of variables present in the dashboard itself.

The "deadlock" I'm observing when loading the dashboard, immediately after grafana container restart, seems to have a direct relation with the configured `GF_DATAPROXY_TIMEOUT`, it's almost as if the timeout behaves like a "sleep" that just blocks `CloudWatch` variable loading for approximately `4x` the timeout value (the values of the variables get correctly populated for each scenario below):
- timeout=30 - panel takes 2 minutes to load
- timeout=60 - panel takes 4 minutes to load
- timeout=10 - panel takes 40 seconds to load
- timeout=5 - panel takes 20 seconds to load
- timeout=2 - panel takes 8 seconds to load

Panels with non `CloudWatch` query variables load instantly.

Hope it helps.

**Comment by andresmgot** [comment]
Interesting, this is just an assumption but it seems that what it's happening for you is that the query that you are using to retrieve one of the dashboard variables is the troublesome one. The panels won't start loading until the dashboard variables finish loading (AFAIK). This means that the faster that the variable query times out, the sooner your panels will render. If you go to the Network tab of your browser, are you able to see which query is timing out?

**Comment by salavessa** [comment]
There are no timeouts in the browser, it's just the server side which seems to get "stale" during around `4x` the configured `GF_DATAPROXY_TIMEOUT` on Dashboards which contain variables with CloudWatch queries only, all other grafana pages work just fine.
This doesn't happen only immediately after grafana container restarts, it also happens while grafana is running as usual but I'm not able to reproduce at will (sometimes the dashboard loads instantly sometimes it "stales"), the restart thing is just a way to always be able to reproduce it.
Also, it seems a global lock as if I open a second tab half way through the load of the first "stale" one, they all end up loading the dashboard at the same time.

**Comment by salavessa** [comment]
Below are some steps where I was able to reproduce the issue in a vanilla `grafana:8.4.3` image running docker in an EC2 instance using a role with the `CloudWatch` permissions described @ https://grafana.com/docs/grafana/latest/datasources/aws-cloudwatch/#metrics-only-example:
1. ssh into EC2-instance + create local tunnel in port 3000 (e.g. `ssh -L 3000:127.0.0.1:3000 <EC2-HOST>`)
1. start grafana: `docker run -d --name gg --restart=always -p 3000:3000 -e GF_DATAPROXY_TIMEOUT=5 grafana/grafana:8.4.3`
1. import _CloudWatch_ datasource: `curl -XPOST http://admin:admin@127.0.0.1:3000/api/datasources       -HContent-Type:application/json --data-binary '{"id":1,"uid":"lKvjptL7k","orgId":1,"name":"CloudWatch","type":"cloudwatch","typeName":"CloudWatch","typeLogoUrl":"public/app/plugins/datasource/cloudwatch/img/amazon-web-services.png","access":"proxy","url":"","password":"","user":"","database":"","basicAuth":false,"isDefault":true,"jsonData":{"authType":"default","defaultRegion":"eu-west-1"},"readOnly":false}'`
1. import _cw_ dashboard: `curl -XPOST http://admin:admin@127.0.0.1:3000/api/dashboards/import -HContent-Type:application/json --data-binary '{"dashboard":{"annotations":{"list":[{"builtIn":1,"datasource":"-- Grafana --","enable":true,"hide":true,"iconColor":"rgba(0, 211, 255, 1)","name":"Annotations & Alerts","target":{"limit":100,"matchAny":false,"tags":[],"type":"dashboard"},"type":"dashboard"}]},"editable":true,"fiscalYearStartMonth":0,"graphTooltip":0,"id":null,"iteration":0,"links":[],"liveNow":false,"panels":[],"schemaVersion":35,"style":"dark","tags":[],"templating":{"list":[{"current":{"selected":false,"text":"All","value":"$__all"},"definition":"dimension_values(us-west-1,AWS/ApplicationELB,ConsumedLCUs,LoadBalancer)","hide":0,"includeAll":true,"multi":true,"name":"cw1","options":[],"query":"dimension_values(us-west-1,AWS/ApplicationELB,ConsumedLCUs,LoadBalancer)","refresh":1,"regex":"","skipUrlSync":false,"sort":0,"type":"query"}]},"time":{"from":"now-6h","to":"now"},"timepicker":{},"timezone":"","title":"CloudWatch Vars","uid":"cw","version":1,"weekStart":""},"overwrite":true}'`
1. _LOCALLY_: access imported grafana dashboard from local browser using tunnel created initially on http://127.0.0.1:3000/d/cw/
1. restart grafana container: `docker restart gg`
1. _LOCALLY_: refresh browser page as soon as the container gets restarted and observe that it takes approximately `4x` the `GF_DATAPROXY_TIMEOUT` value for the Dashboard Services to complete loading

Hopefully you'll also be able to reproduce the issue on your side using the steps above.
Thanks!

**Comment by marefr** [comment]
Thanks. I hope we should be able to reproduce with that information :+1: 

So maybe it's not recommended to override the AWS SDK HTTP client as we did in #44036 or we're doing it wrongly? Reading this I don't get the impression that they don't recommend overriding the HTTP client, https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/custom-http.html. 

One thing they do include is this which we don't have in Grafana for HTTP transport:
```go
// So client makes HTTP/2 requests
err := http2.ConfigureTransport(tr)
if err != nil {
  return &client, err
}
```

**Comment by andresmgot** [comment]
The troublesome HTTP client was reverted here: https://github.com/grafana/grafana/pull/46370 so this can be closed. We will take the info in this thread into account once we resume the related work.

The fix will be available in 8.4.4 (or you can try a [`main` build](https://hub.docker.com/r/grafana/grafana/tags)). Thanks all for the input!
